### ISSUE TITLE
[serverless project init] does not populate [s-variables-common.json]

### ISSUE THREAD

**Issue by BerndWessels** [serverless/serverless#1043]
Here is the situation:

`s-resource-cf.json` defines an `IAM Role` and an `Output Variable` called `myLambdaRole`.

`s-function.json` needs that variable `"customRole": "${myLambdaRole}",`

BUT if somebody else get my `serverless project` via `git clone` he doesn't get the `_meta` folder.

Now he calls `serverless project init` with the same stage and region. That creates the `_meta` folder BUT it does NOT populate the `s-variables-common.json` with the `Output Variables` from `s-resources-cf.json`.

Now he tries to deploy with `serverless dash deploy` and that errors

    Serverless: WARNING: This variable is not defined: myLambdaRole

Unfortunately even calling `serverless resources deploy` will not fix the problem because it says

    Serverless: Deploying resources to stage "dev" in region "us-east-1" via Cloudformation (~3 minutes)...  
    Serverless: No resource updates are to be performed.  

and the `s-variables-common.json` is still not populated with the necessary output variables.

What that means is basically that it is impossible to work as a team together at the same stage in the same region with the same resources when sharing the project via `Git`.

So since we don't want to check in the `_meta` folder into `Git` I would suggest that a `serverless project init` call should make sure that all the `Output Variables` are properly fetched and populated in the `s-variables-common.json`.

This is pretty important, or how do you guys share projects via 'Git' ?

**Comment by eahefnawy** [comment]
Thanks for reaching out @BerndWessels ... I'm notsure I fully understand the issue, what do you mean by "populating `s-variables-common.json`"? This is one of the variables files that is used to populate other `s-*.json` files, but this file **it self** can't be populated (it wouldn't make sense to reference a variable inside a variable file)

Am I missing something here?

**Comment by BerndWessels** [comment]
@eahefnawy In other words - when you checkout a project from `git` - it comes without the `_meta` folder . Now it is impossible for you to `serverless dash deploy` successfully when your `Lambda Functions` require `Output Values` from your resources defined in `s-resource-cf.json`.
Which basically makes it impossible for teams to work on the same project sharing via `git`.

**Comment by BerndWessels** [comment]
Oh and I think the priority is high, or can you explain to me a workaround which enables teams to work together on serverless projects with already deployed resources (without manually sharing the `_meta` folder?
At the moment this is a show stopper for us.

**Comment by HyperBrain** [comment]
@BerndWessels @eahefnawy  I like the idea of `project init` getting the output variables directly from the CF stack to initialize the vars per region. Another solution would be, that the meta-sync plugin would do the same thing - because getting the region variables is only half of the way as there might be important variables defined in `...-common` or in the stage file.
Eventually even the `variables` command could support a `populate` subcommand.
So `sls variables populate` would get the stack vars (e.g. on a build server in a CI environment).
I already fixed the variables command to support the different hierarchies of vars (pending PR #1002 ) and I could easily add the `populate` in a second step.

**Comment by BerndWessels** [comment]
@HyperBrain That would be great. As I said, at the moment we don't see a way to progress without it. Let me know how it goes, thanks.

**Comment by JeremyFrederick** [comment]
The way our team has handled it is that each developer must set up their own stage (ie stage = jfrederick). That way the variables are created for them when they project init on their local machine. The only issue we have is if we have custom variables in s-variables-stage-region.json, in which case we have to manually add them to the _meta/variables folder before running project init.

**Comment by BerndWessels** [comment]
Thanks @JeremyFrederick . But that creates a huge amount of resources and with about 200 developers that is not really an option for us.
We basically have only one dev and one prod environment per customer which is based on the master and dev branch in Git and every developer should be able to work on it at any time.
For now we are trying the `serverless-meta-sync` plugin. Not sure yet if we have to fork it to make it fit for our needs, but it is an option I guess.

**Comment by flomotlik** [comment]
@BerndWessels for the next version of Serverless we'd love to completely remove the meta folder and allow you two things:

1. You can put variables that are not security related directly into the repo
2. Security related keys (e.g. API Keys, ...) can be stored encrypted in your repository
3. **OPTIONAL** Security related keys are stored in a dynamodb encrypted and there is only a key in the repo to know which variable to load (can even be encrypted)

**Comment by BerndWessels** [comment]
@flomotlik That sounds great - being completely CloudFormation based and getting rid of the _meta folder would make things so much cleaner and easier. Looking forward to it.

**Comment by Josh-a-e** [comment]
Hey, I'm running into a related issue (a lack of variables in my _meta folder) but in the context of "I need to deploy this stuff via CI".

What's happening is:
- commit gets made to master
- CI starts up, runs `sls project init` to create a basic _meta folder, but void of most of the details I require
- I run `sls resources deploy`, but more often than not there's no cloudformation changed to be made, and I see a "No resource updates are to be performed." message: https://github.com/serverless/serverless/blob/078a0281183cf1c376b2e95994326d8ba1ce2b5c/lib/actions/ResourcesDeploy.js#L306 Fair enough, but actually I'm now left with a _meta folding lacking the variables output by a "successful" (i.e. with a change) `sls resources deploy` call.
- My application now fails to deploy

My temporary solution is using meta-sync, but this in turn is hamstrung by its inability to write to the s3 bucket whilst in CI mode: https://github.com/serverless/serverless-meta-sync/blob/master/lib/MetaSync.js#L159

So in the case where actually there _is_ a change to cloudformation and an output changes, this new variable value isn't getting added back to the bucket, so actually deploying cloudformation changes is only doable from my local machine with the master credentials.

(Side note: In a large organisation (e.g. that of @BerndWessels), I can certainly imagine a case where a developer doesn't have permission to make such changes to production cf templates, so submitting a PR with the altered cloudformation template for review would be the logical, "correct" step and my alternative isn't an option)

My actual suggestion here is simply to make a call to `describeStacks` (http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#describeStacks-property) when the "No resource updates are to be performed" outcome is detected (linked above). This returns the outputs which can then be used to go ahead and populate the _meta folder as desired.

I'd be really interested in helping if I can, although it sounds like you guys might have work in the pipeline that makes some / all of this redundant.

Thanks for your time!

**Comment by BerndWessels** [comment]
@Josh-a-e Great points, especially the one about PR's to change resources.

We are also experimenting with the correct CI approach with AWS and serverless and the _meta folder is just not working.

I hope that moving to a complete CloudFormation based solutions without _meta folder as @flomotlik mentioned might solve all these problems.

I would also be keen on contributing if needed.

**Comment by c-knowles** [comment]
+1 on CloudFormation as that would likely tidy this up a lot. One slight downside is I think it's probably going to be slower to deploy, at least that's been my experience of CF versus doing the CLI calls directly.

Also +1 on some way to initialise the output variables of a stage.

I thought I'd add my current strategy for CD on a project using sls v0.5.5 and codeship. It's definitely more of a struggle than it should be with the current setup but maybe this will help someone trying similar before v1 is out. What I ended up with boils down to:
* Committing `_meta/variables`
* Codeship uses a Docker container of your own definition to build/deploy. Ensure `_meta/variables` gets moved elsewhere so that the container has no `_meta/variables` folder but still has access to the contents (let's say `serverless_meta/variables`). This part ensures serverless doesn't complain about the environment already existing as it sometimes does.
* Run `sls project init -n PROJECT_NAME -r REGION -s ENV` which ensures the target environment is up to date. We also deploy with a stage equal to the git hash at some point in our pipeline so we want to setup the environment in this case too.
* After this, if there is a match for the deploy environment in the `serverless_meta/variables`, copy the stage and region files into `_meta/variables`. This ensures any more non-secret variables we populated specific to an environment are setup correctly.
* Do a few calls to `sls variables set -k KEY -v VALUE -s ENV -r REGION` to populate some secret non-committed variables.
* Finally run `sls function deploy -a -s ENV -r REGION` followed by `sls endpoint deploy -a -s ENV -r REGION` (we don't use the events that much but easy enough to add it).

I suppose the main difference to the above cases is having the `_meta` folder committed but moving it for CI so serverless can fully initialise itself. I haven't had to run `sls resources deploy`.

The above does not help with team collaboration and I could remove the whole copying of the variable files if I setup all variables via `sls variables set` although that would work directly against team collaboration since then everyone would have to run various additional scripts to populate the variables rather than just having the `sls` CLI work on project checkout.
